orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,1365 @@
+                return
+
+        item = QStandardItem()
+        item.setData({}, ParametersRole)
+        item.setData(pp_action.title, Qt.DisplayRole)
+        item.setData(pp_action, DescriptionRole)
+        self.preprocessormodel.appendRow([item])
+
+    def storeSpecificSettings(self):
+        self.stored_settings = self.save(self.preprocessormodel)
+        super().storeSpecificSettings()
+
+    def commit(self):
+        # Sync the model into stored_settings on every change commit.
+        self.storeSpecificSettings()
+        if not self._invalidated:
+            self._invalidated = True
+            QApplication.postEvent(self, QEvent(QEvent.User))
+
+    def handleNewSignals(self):
+        self.apply()
+
+    def customEvent(self, event):
+        if event.type() == QEvent.User and self._invalidated:
+            self._invalidated = False
+            self.apply()
 
     def apply(self):
-        # TODO change this to custom stopwords
-        if self.remove_stpwrds:
-            sw = 'english'
-        pp = Preprocessor(incl_punct=self.include_punctuation, trans=self.transformation,
-                          lowercase=self.lowercase, stop_words=sw)
-        self.send(Output.PREPROCESSOR, pp)
\ No newline at end of file
+        self.error(1, "")
+        # Nothing to do, if there is no input corpus.
+        if self.data is None:
+            return
+
+        # Build a preprocessor instance with the selected parameters.
+        preprocessor = self.build_preproc()
+        if not preprocessor:
+            self.error(1, "No tokenizer specified.")
+            return
+
+        # Take the input corpus and append tokens to the metadata.
+        input_data = self.data.documents
+        output_corp = self.data
+        tokens = preprocessor(input_data)
+
+        print(tokens)
+
+        # Update the input corpus with a list of tokens.
+        self.data.tokens = tokens
+        self.send(Output.PP_CORPUS, self.data)
+
+        # Sync the model into stored_settings on every apply.
+        self.storeSpecificSettings()
+
+    def build_preproc(self):
+        self.warning(1, "")
+        # Preprocessor parameters that we require.
+        p = {"tokenizer": None, "lowercase": False, "stop_words": None, "transformation": None}
+        # For each row in the self.preprocessormodel, acquire
+        # the info for the preprocessor.
+        for i in range(self.preprocessormodel.rowCount()):
+            pp_item = self.preprocessormodel.item(i)
+            desc = pp_item.data(DescriptionRole)
+            params = pp_item.data(ParametersRole)
+
+            if not isinstance(params, dict):
+                params = {}
+
+            # Get the method from the BaseEditor class.
+            pp_info = desc.view_class.create_instance(params)
+
+            # Custom sw file. Read into list.
+            if "stop_words" in pp_info and pp_info["stop_words"] not in ["default", "None"]:
```

Comment:
Reading stopword files shoud be done elsewhere. Now you read it even if some unrelated settings were modified. This shoud be done imediatly after the user selects the file from the dropdown on file system.